Improve checksum checks for async INSERT into Distributed on the sender

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Check per-block checksum of the distributed batch on the sender before sending (without reading the file twice, the checksums will be verified while reading), this will avoid stuck of the INSERT on the receiver (on truncated .bin file on the sender)
- Avoid reading .bin files twice for batched INSERT (it was required to calculate rows/bytes to take squashing into account, now this information included into the header, backward compatible is preserved).

Refs: #18369

azat committed:
Add a test for corruption during distributed send

azat committed:
Check per-block checksum of the distributed batch on the sender before sending

This is already done for distributed_directory_monitor_batch_inserts=1,
so let's do the same for the non batched mode, since otherwise in case
the file will be truncated the receiver will just stuck (since it will
wait for the block, but the sender will not send it).

azat commented on src/Storages/Distributed/DirectoryMonitor.cpp:
```diff
@@ -311,6 +311,35 @@ void StorageDistributedDirectoryMonitor::processFile(const std::string & file_pa
     auto timeouts = ConnectionTimeouts::getTCPTimeoutsWithFailover(storage.global_context.getSettingsRef());
 
+    /// Check that the file is valid
+    /// FIXME: suboptimal
```
Another way is to store the size of the compressed file and number of rows in it in the DistributedBlockOutputStream::writeToShard, and then decompression code path can be removed completely (not only this place but also processFilesWithBatches) from the DirectoryMonitory

alexey-milovidov replied:
Yes, we need to check it without decompression.
CheckingCompressedReadBuffer will be Ok to implement.
The format was designed to allow sending pre-compressed data.

azat replied:
But it will still require calculating hashes, so reading the whole file.

I was thinking about adding file size to the header of the .bin file and verify only it, thus we can avoid reading the whole file.
Since we only need to protect from the truncation on the sender, the receiver will check per-compressed block checksums anyway and fail with an error.

filimonov replied:
Yeah, may be we can store new format as xxx.bin2 and use that double-read approach with unpacking / checksum checking for old format (bin).

alexey-milovidov replied:
> But it will still require calculating hashes, so reading the whole file.

Maybe do it while sending?

azat replied:
>Maybe do it while sending?

@alexey-milovidov the receiver will check per-block checksum anyway, are you suggesting checking them on the sender? This will complicated the code, and I'm not sure that this worth it.

azat replied:
>Yeah, may be we can store new format as xxx.bin2 and use that double-read approach with unpacking / checksum checking for old format (bin).

@filimonov The format already have the header, no need to introduce another extension.

alexey-milovidov replied:
> the receiver will check per-block checksum anyway, are you suggesting checking them on the sender?

Yes.

> This will complicated the code, and I'm not sure that this worth it.

Then it's not necessarily.
But we should not read data twice.

If you can check the size it will be enough.

filimonov replied:
> @filimonov The format already have the header, no need to introduce another extension

Backward  compatibility and UX during version upgrade concerns: Will both new a old versions will know how to read files with old/new header? BTW there is also File format to read files from Distributed folder AFAIR.

azat replied:
>CheckingCompressedReadBuffer will be Ok to implement.

@alexey-milovidov I went this way, since size of the file cannot be stored in the header (without trickery), although it can be stored in the footer, but I decided that implementing CheckingCompressedReadBuffer will be better (it will verify checksums and avoid seekign back and forth in the file, since after seeking to `file_size-sizeof(footer)` you need to seek back or open two file descriptors).

azat replied:
>Backward compatibility and UX during version upgrade concerns: Will both new and old versions know how to read files with old/new header?

@filimonov Yes

Current version changes the on disk format, but not for checksums, but to include rows/bytes to avoid reading the full file to obtain this information (it is required to do proper batching), and if this information does not exists in the header it will be calculated as before (by reading whole file).

>BTW there is also File format to read files from Distributed folder AFAIR.

It uses the same code so it should be fully compatible.

azat committed:
Refactoring distributed header parsing

azat committed:
Add more metadata into distributed .bin files to avoid doing the same on sending

Before this patch StorageDistributedDirectoryMonitor reading .bin files
in batch mode, just to calculate number of bytes/rows, this is very
ineffective, let's just store them in the header (rows/bytes).

azat committed:
Verify compressed headers while sending distributed batches

Before this patch the DirectoryMonitor was checking the compressed file
by reading it one more time (since w/o this receiver may stuck on
truncated file), while this is ineffective and we can just check the
checksums before sending.

But note that this may decrease batch size that is used for sending over
network.

azat committed:
Add CheckingCompressedReadBuffer

Buffer for reading from a compressed file with just checking checksums
of the compressed blocks, without any decompression, so result can be
proxied.

azat committed:
Fix readability-static-definition-in-anonymous-namespace in DirectoryMonitor

alexey-milovidov commented:
https://github.com/ClickHouse/ClickHouse/issues/18887

azat committed:
Update test_check_table expectations

Since now CompressedReadBufferBase read checksum and header at once it
will expect 25 (16+9) bytes not 16 (only checksum)

azat commented on src/Storages/Distributed/DirectoryMonitor.cpp:
```diff
@@ -715,28 +731,45 @@ void StorageDistributedDirectoryMonitor::processFilesWithBatching(const std::map
+                total_bytes += header.bytes;
 
-                if (!sample_block)
+                CompressedReadBuffer decompressing_in(in);
+                NativeBlockInputStream block_in(decompressing_in, DBMS_TCP_PROTOCOL_VERSION);
+                block_in.readPrefix();
+
+                /// We still need to read one block for the header.
+                while (Block block = block_in.read())
+                {
                     sample_block = block.cloneEmpty();
+                    break;
+                }
```
In c939d2f96ee71beaffbf41429080078c3d206ece Distributed INSERTs was splitted by block structure, and this requires reading the whole file (due to the nature of NativeBlockInputStream)

@alexey-milovidov maybe you can provide some details for CLICKHOUSE-3163 (it was referenced in the mentioned commit)?

azat replied:
Actually maybe it was required before #10135 and since it was merged I don't see any reason for doing this

azat replied:
Nevermind, this is still relevant, since the underlying table can be modified while distributed not, and in this case the whole batch will fail, need better approach to avoid extra file reading.

azat replied:
Should be addressed in ecae6c1c609fbabaf940ec41c40c7f3b2401b412

azat replied:
And let's cover this behavior - #19155

azat commented:
Here are some numbers, in general, they shows that `CheckingCompressedReadBuffer` has almost zero overhead.

### patched

```sql
create table data (key Int, value String) engine=Null();
create table dist (key Int, value String) engine=Distributed(test_shard_localhost, currentDatabase(), data);
system stop distributed sends dist;

-- insert into Null() directly, just foward blocks to Null engine.
insert into dist select number, reinterpretAsString(number) from system.numbers limit 1e9 settings prefer_localhost_replica=1;
-- 0 rows in set. Elapsed: 16.888 sec. Processed 1.00 billion rows, 8.00 GB (59.23 million rows/s., 473.85 MB/s.)

-- async INSERT, with storing blocks on disk
insert into dist select number, reinterpretAsString(number) from system.numbers limit 1e9 settings prefer_localhost_replica=0;
-- 0 rows in set. Elapsed: 44.010 sec. Processed 1.00 billion rows, 8.00 GB (22.73 million rows/s., 181.83 MB/s.)

-- distributed_directory_monitor_batch_inserts=0
system flush distributed dist;
-- pending *.bin size is 7.5G
-- 0 rows in set. Elapsed: 22.445 sec.

-- distributed_directory_monitor_batch_inserts=1
system flush distributed dist;
-- first run
-- 0 rows in set. Elapsed: 50.845 sec.;
-- second run
-- 0 rows in set. Elapsed: 39.750 sec.;
```

### upstream

```sql
-- distributed_directory_monitor_batch_inserts=0
system flush distributed dist;
-- 0 rows in set. Elapsed: 21.459 sec.

-- distributed_directory_monitor_batch_inserts=1
system flush distributed dist;
-- 0 rows in set. Elapsed: 45.488 sec.
```

I believe that batching mode overhead can be eliminated by removing extra file reading.

*NOTE: INSERT was done using the same version as `system flush distributed dist`, this is required to fill metadata correctly*

azat commented:
>And I believe that batching mode overhead can be eliminated by removing extra file reading.

Confirmed, w/o splitting by block header it is 2x faster (since does not requires extra file reading after other patches in this series)

azat committed:
Avoid reading the distributed batch just to read the block header

Before this patch batched mode of the DirectoryMonitor is 2x slower then
non-batched, after it should be more or less the same as non-batched.

azat commented:
>AST fuzzer — Logical error: 'Block structure mismatch in PipelineExecuting stream: different columns:

#19007
